Add API to temporalily suppress usage of ASAN's fake stack

happyCoder92 opened the pull request:
Intended use-case is for threads that use (or switch to) stack with special properties e.g. backed by MADV_DONTDUMP memory.

happyCoder92 committed:
Add API to temporalily disable usage of ASAN's fake stack

Intended use-case is for threads that use (or switch to) stack
with special properties e.g. backed by MADV_DONTDUMP memory.

happyCoder92 committed:
Add API to temporalily disable usage of ASAN's fake stack

Intended use-case is for threads that use (or switch to) stack
with special properties e.g. backed by MADV_DONTDUMP memory.

vitalybuka commented on compiler-rt/include/sanitizer/asan_interface.h:
```diff
@@ -333,6 +333,13 @@ void SANITIZER_CDECL __asan_handle_no_return(void);
 int SANITIZER_CDECL __asan_update_allocation_context(void *addr);
 
+/// Disables fake stack usage for the current thread.
+/// Temporarily disables use-after-return detection for current thread.
```
Please remove the word ""usage" from the comment and function name.
Seems redundunt.

vitalybuka reviewed (changes requested):
comments above

vitalybuka commented on compiler-rt/lib/asan/asan_fake_stack.cpp:
```diff
@@ -217,26 +217,44 @@ static THREADLOCAL FakeStack *fake_stack_tls;
+  }
+}
+
+static FakeStack* GetFakeStack(bool for_allocation = true) {
```
I am not sure understand what do we need for_allocation at all?

happyCoder92 replied:
Removed the argument.

vitalybuka commented on compiler-rt/lib/asan/asan_fake_stack.cpp:
```diff
@@ -311,7 +329,9 @@ extern "C" {
 // (https://github.com/google/sanitizers/issues/1394)
 SANITIZER_INTERFACE_ATTRIBUTE
-void *__asan_get_current_fake_stack() { return GetFakeStackFast(); }
+void* __asan_get_current_fake_stack() {
```
There is TODO for to remove this one
if this is just for test, can we instead rely on __asan_get_current_fake_stack
Use __builtin_frame_address

__builtin_frame_address always return real stack address

happyCoder92 replied:
Removed the argument and adjusted the test.

vitalybuka replied:
Did you forget to upload the patch?

vitalybuka replied:
I have rebased and updated it

vitalybuka commented on compiler-rt/lib/asan/asan_fake_stack.cpp:
```diff
@@ -217,26 +217,44 @@ static THREADLOCAL FakeStack *fake_stack_tls;
   return fake_stack_tls;
 }
-void SetTLSFakeStack(FakeStack *fs) {
+void SetTLSFakeStack(AsanThread* t, FakeStack* fs) {
```
Will #163481 help to simplify this patch?

happyCoder92 replied:
Slightly. By only calling `SetTLSFakeStack` from `GetFakeStack` #163481 also indirectly ensures the tls var won't be set if fake stack is disabled. I added the thread arg and put the logic into `SetTLSFakeStack` to make it more explicit.

happyCoder92 commented on compiler-rt/lib/asan/asan_fake_stack.cpp:
```diff
@@ -217,26 +217,44 @@ static THREADLOCAL FakeStack *fake_stack_tls;
+
+static FakeStack* GetFakeStack(bool for_allocation = true) {
+  AsanThread* t = GetCurrentThread();
+  if (!t || (for_allocation && !t->IsFakeStackEnabled()))
```
Removed the argument.

vitalybuka committed:
Revert "Add API to temporalily disable usage of ASAN's fake stack"

This reverts commit 4e728017929bfa33edb2bd0a6c82f8f25468c22c.

vitalybuka committed:
Merge branch 'happyCoder92/main' into HEAD

vitalybuka committed:
Update test

vitalybuka committed:
format

vitalybuka committed:
format

vitalybuka committed:
format

thurstond reviewed (changes requested):
What is the intended interaction of this API with `-mllvm -asan-use-after-return={never,runtime,always}` and `ASAN_OPTIONS=detect_stack_use_after_return={0,1}`? Please update the description and `TestCases/disable_fake_stack.cpp` to clarify.

Also, if the behavior of `__asan_disable_fake_stack()`/`__asan_enable_fake_stack()` depends on those flags, the naming could be confusing.

vitalybuka commented:
> The behavior does depend on the flags. Would `__asan_suppress_fake_stack` and `__asan_unsuppress_fake_stack` be less confusing?

I don't have a preference for naming.

I assume @thurstond  is asking not to disable the fake stack in always mode.

But in this case __asan_disable_fake_stack caller needs to know that the call failed?

I the current `__asan_get_current_fake_stack` can be used for that.

Also we need actually not a bool but counters, see __lsan_disable().

This will fix the issue when code on multiple layers will try to suppress the fake stack.

thurstond commented:
> > For 2., since __asan_disable_fake_stack()/__asan_enable_fake_stack() is a runtime API, it is debatable why it can't
> 
> The point of '=2' is to change Asan pass to generate code which does not check for nullptr.
> 
> Regarding documentation, In this header yes, in Clang docs, better not yet.

By "For 2.", I was referring to bullet point 2 (sorry for the ambiguity):

2. compiled with "runtime", ASAN_OPTIONS=detect_stack_use_after_return=0: fake stack is always off

Why can't this be overridden by  `__asan_disable_fake_stack()`/`__asan_enable_fake_stack()`?

happyCoder92 committed:
Add thorough tests

happyCoder92 committed:
Rename the functions & use counter

happyCoder92 commented:
I added tests to clarify the behavior.
Changed the naming and also used a counter.

As stated before the intention is to always suppress/disable fake stack `__asan_suppress_fake_stack()`, but not reenable it with `__asan_unsuppress_fake_stack()` if disabled via some other option.

happyCoder92 committed:
Code formatting

happyCoder92 committed:
Merge branch 'main' into main

happyCoder92 committed:
Rename test utility & try to fix it for Windows

happyCoder92 committed:
Explicitly set UAR option as those default to false on Windows

happyCoder92 commented on compiler-rt/test/asan/TestCases/suppress_fake_stack.cpp:
```diff
@@ -0,0 +1,34 @@
+  return this_frame <= saved && saved <= caller_frame;
+}
+
+ATTRIBUTE_NOINLINE bool IsOnStack() {
```
Done.

happyCoder92 committed:
Merge branch 'main' into main
